File: alligator/src/main/java/me/aartikov/alligator/AndroidNavigator.java
package me.aartikov.alligator;

import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.LinkedList;
import java.util.Queue;

import me.aartikov.alligator.animations.AnimationData;
import me.aartikov.alligator.commands.BackCommand;
import me.aartikov.alligator.commands.BackToClassCommand;
import me.aartikov.alligator.commands.BackToScreenCommand;
import me.aartikov.alligator.commands.Command;
import me.aartikov.alligator.commands.FinishCommand;
import me.aartikov.alligator.commands.ForwardCommand;
import me.aartikov.alligator.commands.ReplaceCommand;
import me.aartikov.alligator.commands.ResetCommand;
import me.aartikov.alligator.commands.SwitchToCommand;
import me.aartikov.alligator.exceptions.NavigationException;
import me.aartikov.alligator.navigationfactories.NavigationFactory;


/**
 * Main library object. It translates calls of navigation methods to commands and puts them to a command queue.
 * Commands can be executed if a {@link NavigationContext} is bound, otherwise command execution will be delayed until a {@link NavigationContext} will be bound.
 */
public class AndroidNavigator implements NavigationContextBinder, Navigator {
    private NavigationFactory mNavigationFactory;
    private NavigationContext mNavigationContext;
    private Queue<Command> mCommandQueue = new LinkedList<>();
    private boolean mIsExecutingCommands;
    private ScreenResolver mScreenResolver;
    private ActivityResultHandler mActivityResultHandler;

    public AndroidNavigator(NavigationFactory navigationFactory) {
        mNavigationFactory = navigationFactory;
        mScreenResolver = new ScreenResolver(navigationFactory);
        mActivityResultHandler = new ActivityResultHandler(navigationFactory);
    }

    public NavigationFactory getNavigationFactory() {
        return mNavigationFactory;
    }

    public ScreenResolver getScreenResolver() {
        return mScreenResolver;
    }

    public ActivityResultHandler getActivityResultHandler() {
        return mActivityResultHandler;
    }

    @Nullable
    @Override
    public NavigationContext getNavigationContext() {
        return mNavigationContext;
    }

    @Override
    public boolean isBound() {
        return mNavigationContext != null;
    }

    @Override
    public void bind(@NonNull NavigationContext navigationContext) {
        checkThatMainThread();
        if (mNavigationContext != null && mNavigationContext.getActivity() != navigationContext.getActivity()) {
            return;
        }
        mNavigationContext = navigationContext;
        mActivityResultHandler.setScreenResultListener(mNavigationContext.getScreenResultListener());
        executeQueuedCommands();
    }

    @Override
    public void unbind(AppCompatActivity activity) {
        checkThatMainThread();
        if (mNavigationContext != null && mNavigationContext.getActivity() != activity) {
            return;
        }
        mActivityResultHandler.resetScreenResultListener();
        mNavigationContext = null;
    }

    @Override
    public boolean canExecuteCommandImmediately() {
        return mCommandQueue.isEmpty() && mNavigationContext != null;
    }

    @Override
    public boolean hasPendingCommands() {
        return !mCommandQueue.isEmpty();
    }

    /**
     * Adds a new screen and goes to it. Implemented with {@link ForwardCommand}.
     *
     * @param screen new screen
     */
    @Override
    public void goForward(@NonNull Screen screen) {
        goForward(screen, null);
    }

    @Override
    public void goForward(@NonNull Screen screen, @Nullable AnimationData animationData) {
        executeCommand(new ForwardCommand(screen, animationData));
    }

    /**
     * Finishes a current screen and goes back to the previous screen. Implemented with {@link BackCommand}.
     */
    @Override
    public void goBack() {
        goBack(null);
    }

    @Override
    public void goBack(@Nullable AnimationData animationData) {
        executeCommand(new BackCommand(null, animationData));
    }

    /**
     * Finishes a current screen and goes back to the previous screen with result. Implemented with {@link BackCommand}.
     */
    @Override
    public void goBackWithResult(@NonNull ScreenResult screenResult) {
        goBackWithResult(screenResult, null);
    }

    @Override
    public void goBackWithResult(@NonNull ScreenResult screenResult, @Nullable AnimationData animationData) {
        executeCommand(new BackCommand(screenResult, animationData));
    }

    /**
     * Goes back to a screen with the given class. Implemented with {@link BackToClassCommand}.
     *
     * @param screenClass screen class for going back
     */
    @Override
    public void goBackTo(@NonNull Class<? extends Screen> screenClass) {
        goBackTo(screenClass, null);
    }

    /**
     * Goes back to a screen with the given screen object. Implemented with {@link BackToScreenCommand}.
     *
     * @param screen screen for going back
     */
    @Override
    public void goBackTo(@NonNull Screen screen) {
        goBackTo(screen, null);
    }

    @Override
    public void goBackTo(@NonNull Class<? extends Screen> screenClass, @Nullable AnimationData animationData) {
        executeCommand(new BackToClassCommand(screenClass, null, animationData));
    }

    @Override
    public void goBackTo(@NonNull Screen screen, @Nullable AnimationData animationData) {
        executeCommand(new BackToScreenCommand(screen, null, animationData));
    }

    /**
     * Goes back to a screen with the given class and returns result to it. Implemented with {@link BackToClassCommand}.
     *
     * @param screenClass  screen class for going back
     * @param screenResult screen result that will be returned
     */
    @Override
    public void goBackToWithResult(@NonNull Class<? extends Screen> screenClass, @NonNull ScreenResult screenResult) {
        goBackToWithResult(screenClass, screenResult, null);
    }

    /**
     * Goes back to a screen with the given object screen and returns result to it. Implemented with {@link BackToScreenCommand}.
     *
     * @param screen  screen class for going back
     * @param screenResult screen result that will be returned
     */
    @Override
    public void goBackToWithResult(@NonNull Screen screen, @NonNull ScreenResult screenResult) {
        goBackToWithResult(screen, screenResult, null);
    }

    @Override
    public void goBackToWithResult(@NonNull Class<? extends Screen> screenClass, @NonNull ScreenResult screenResult, @Nullable AnimationData animationData) {
        executeCommand(new BackToClassCommand(screenClass, screenResult, animationData));
    }

    @Override
    public void goBackToWithResult(@NonNull Screen screen, @NonNull ScreenResult screenResult, @Nullable AnimationData animationData) {
        executeCommand(new BackToScreenCommand(screen, screenResult, animationData));
    }

    /**
     * Replaces the last screen with a new screen. Implemented with {@link ReplaceCommand}.
     *
     * @param screen new screen
     */
    @Override
    public void replace(@NonNull Screen screen) {
        replace(screen, null);
    }

    @Override
    public void replace(@NonNull Screen screen, @Nullable AnimationData animationData) {
        executeCommand(new ReplaceCommand(screen, animationData));
    }

    /**
     * Removes all other screens and adds a new screen. Implemented with {@link ResetCommand}.
     *
     * @param screen new screen
     */
    @Override
    public void reset(@NonNull Screen screen) {
        reset(screen, null);
    }

    @Override
    public void reset(@NonNull Screen screen, @Nullable AnimationData animationData) {
        executeCommand(new ResetCommand(screen, animationData));
    }

    /**
     * Finishes a current flow or a current top-level screen (represented by activity). Implemented with {@link FinishCommand}.
     */
    @Override
    public void finish() {
        finish(null);
    }

    @Override
    public void finish(@Nullable AnimationData animationData) {
        executeCommand(new FinishCommand(null, false, animationData));
    }

    /**
     * Finishes a current flow or a current top-level screen (represented by activity) with result. Implemented with {@link FinishCommand}.
     *
     * @param screenResult screen result that will be returned
     */
    @Override
    public void finishWithResult(@NonNull ScreenResult screenResult) {
        finishWithResult(screenResult, null);
    }

    @Override
    public void finishWithResult(@NonNull ScreenResult screenResult, @Nullable AnimationData animationData) {
        executeCommand(new FinishCommand(screenResult, false, animationData));
    }

    /**
     * Finishes a current top-level screen (represented by activity). Implemented with {@link FinishCommand}.
     */
    @Override
    public void finishTopLevel() {
        finishTopLevel(null);
    }

    @Override
    public void finishTopLevel(@Nullable AnimationData animationData) {
        executeCommand(new FinishCommand(null, true, animationData));
    }

    /**
     * Finishes a current top-level screen (represented by activity) with result. Implemented with {@link FinishCommand}.
     *
     * @param screenResult screen result that will be returned
     */
    @Override
    public void finishTopLevelWithResult(@NonNull ScreenResult screenResult) {
        finishTopLevelWithResult(screenResult, null);
    }

    @Override
    public void finishTopLevelWithResult(@NonNull ScreenResult screenResult, @Nullable AnimationData animationData) {
        executeCommand(new FinishCommand(screenResult, true, animationData));
    }

    /**
     * Switches screen with a screen switcher. Implemented with {@link SwitchToCommand}.
     *
     * @param screen screen
     */
    @Override
    public void switchTo(@NonNull Screen screen) {
        switchTo(screen, null);
    }

    @Override
    public void switchTo(@NonNull Screen screen, AnimationData animationData) {
        executeCommand(new SwitchToCommand(screen, animationData));
    }

    protected void executeCommand(Command command) {
        checkThatMainThread();
        mCommandQueue.add(command);
        executeQueuedCommands();
    }

    private void executeQueuedCommands() {
        if (mIsExecutingCommands) {
            return;
        }

        mIsExecutingCommands = true;
        try {
            while (mNavigationContext != null && !mCommandQueue.isEmpty()) {
                Command command = mCommandQueue.remove();
                boolean canExecuteCommands = command.execute(mNavigationContext);
                if (!canExecuteCommands) {
                    mNavigationContext = null;
                }
            }
        } catch (NavigationException e) {
            mCommandQueue.clear();
            mNavigationContext.getNavigationErrorListener().onNavigationError(e);
        } catch (Exception e) {
            mCommandQueue.clear();
            throw e;
        } finally {
            mIsExecutingCommands = false;
        }
    }

    private void checkThatMainThread() {
        if (Thread.currentThread() != Looper.getMainLooper().getThread()) {
            throw new RuntimeException("Can only be called from the main thread.");
        }
    }
}

File: alligator/src/main/java/me/aartikov/alligator/AndroidNavigator.kt
package me.aartikov.alligator

import android.os.Looper
import androidx.appcompat.app.AppCompatActivity
import me.aartikov.alligator.animations.AnimationData
import me.aartikov.alligator.commands.BackCommand
import me.aartikov.alligator.commands.BackToClassCommand
import me.aartikov.alligator.commands.BackToScreenCommand
import me.aartikov.alligator.commands.Command
import me.aartikov.alligator.commands.FinishCommand
import me.aartikov.alligator.commands.ForwardCommand
import me.aartikov.alligator.commands.ReplaceCommand
import me.aartikov.alligator.commands.ResetCommand
import me.aartikov.alligator.commands.SwitchToCommand
import me.aartikov.alligator.exceptions.NavigationException
import me.aartikov.alligator.navigationfactories.NavigationFactory
import java.util.LinkedList
import java.util.Queue

/**
 * Main library object. It translates calls of navigation methods to commands and puts them to a command queue.
 * Commands can be executed if a [NavigationContext] is bound, otherwise command execution will be delayed until a [NavigationContext] will be bound.
 */
class AndroidNavigator(val navigationFactory: NavigationFactory) : NavigationContextBinder, Navigator {

    override var navigationContext: NavigationContext? = null
        private set

    private val commandQueue: Queue<Command> = LinkedList()
    private var isExecutingCommands = false

    val screenResolver: ScreenResolver = ScreenResolver(navigationFactory)
    val activityResultHandler: ActivityResultHandler = ActivityResultHandler(navigationFactory)

    override val isBound: Boolean
        get() = navigationContext != null

    override fun bind(navigationContext: NavigationContext) {
        checkThatMainThread()
        if (this.navigationContext != null
            && this.navigationContext!!.activity !== navigationContext.activity
        ) {
            return
        }

        this.navigationContext = navigationContext
        activityResultHandler.setScreenResultListener(navigationContext.screenResultListener)
        executeQueuedCommands()
    }

    override fun unbind(activity: AppCompatActivity) {
        checkThatMainThread()
        if (navigationContext != null && navigationContext!!.activity !== activity) {
            return
        }

        activityResultHandler.resetScreenResultListener()
        navigationContext = null
    }

    override fun canExecuteCommandImmediately(): Boolean {
        return commandQueue.isEmpty() && navigationContext != null
    }

    override fun hasPendingCommands(): Boolean {
        return !commandQueue.isEmpty()
    }

    /**
     * Adds a new screen and goes to it. Implemented with [ForwardCommand].
     *
     * @param screen new screen
     */
    override fun goForward(screen: Screen) {
        goForward(screen, null)
    }

    override fun goForward(screen: Screen, animationData: AnimationData?) {
        executeCommand(ForwardCommand(screen, animationData))
    }

    /**
     * Finishes a current screen and goes back to the previous screen. Implemented with [BackCommand].
     */
    override fun goBack() {
        goBack(null)
    }

    override fun goBack(animationData: AnimationData?) {
        executeCommand(BackCommand(null, animationData))
    }

    /**
     * Finishes a current screen and goes back to the previous screen with result. Implemented with [BackCommand].
     */
    override fun goBackWithResult(screenResult: ScreenResult) {
        goBackWithResult(screenResult, null)
    }

    override fun goBackWithResult(screenResult: ScreenResult, animationData: AnimationData?) {
        executeCommand(BackCommand(screenResult, animationData))
    }

    /**
     * Goes back to a screen with the given class. Implemented with [BackToClassCommand].
     *
     * @param screenClass screen class for going back
     */
    override fun goBackTo(screenClass: Class<out Screen>) {
        goBackTo(screenClass, null)
    }

    /**
     * Goes back to a screen with the given screen object. Implemented with [BackToScreenCommand].
     *
     * @param screen screen for going back
     */
    override fun goBackTo(screen: Screen) {
        goBackTo(screen, null)
    }

    override fun goBackTo(screenClass: Class<out Screen>, animationData: AnimationData?) {
        executeCommand(BackToClassCommand(screenClass, null, animationData))
    }

    override fun goBackTo(screen: Screen, animationData: AnimationData?) {
        executeCommand(BackToScreenCommand(screen, null, animationData))
    }

    /**
     * Goes back to a screen with the given class and returns result to it. Implemented with [BackToClassCommand].
     *
     * @param screenClass  screen class for going back
     * @param screenResult screen result that will be returned
     */
    override fun goBackToWithResult(screenClass: Class<out Screen>, screenResult: ScreenResult) {
        goBackToWithResult(screenClass, screenResult, null)
    }

    /**
     * Goes back to a screen with the given object screen and returns result to it. Implemented with [BackToScreenCommand].
     *
     * @param screen       screen class for going back
     * @param screenResult screen result that will be returned
     */
    override fun goBackToWithResult(screen: Screen, screenResult: ScreenResult) {
        goBackToWithResult(screen, screenResult, null)
    }

    override fun goBackToWithResult(
        screenClass: Class<out Screen>,
        screenResult: ScreenResult,
        animationData: AnimationData?
    ) {
        executeCommand(BackToClassCommand(screenClass, screenResult, animationData))
    }

    override fun goBackToWithResult(
        screen: Screen,
        screenResult: ScreenResult,
        animationData: AnimationData?
    ) {
        executeCommand(BackToScreenCommand(screen, screenResult, animationData))
    }

    /**
     * Replaces the last screen with a new screen. Implemented with [ReplaceCommand].
     *
     * @param screen new screen
     */
    override fun replace(screen: Screen) {
        replace(screen, null)
    }

    override fun replace(screen: Screen, animationData: AnimationData?) {
        executeCommand(ReplaceCommand(screen, animationData))
    }

    /**
     * Removes all other screens and adds a new screen. Implemented with [ResetCommand].
     *
     * @param screen new screen
     */
    override fun reset(screen: Screen) {
        reset(screen, null)
    }

    override fun reset(screen: Screen, animationData: AnimationData?) {
        executeCommand(ResetCommand(screen, animationData))
    }

    /**
     * Finishes a current flow or a current top-level screen (represented by activity). Implemented with [FinishCommand].
     */
    override fun finish() {
        finish(null)
    }

    override fun finish(animationData: AnimationData?) {
        executeCommand(FinishCommand(null, false, animationData))
    }

    /**
     * Finishes a current flow or a current top-level screen (represented by activity) with result. Implemented with [FinishCommand].
     *
     * @param screenResult screen result that will be returned
     */
    override fun finishWithResult(screenResult: ScreenResult) {
        finishWithResult(screenResult, null)
    }

    override fun finishWithResult(screenResult: ScreenResult, animationData: AnimationData?) {
        executeCommand(FinishCommand(screenResult, false, animationData))
    }

    /**
     * Finishes a current top-level screen (represented by activity). Implemented with [FinishCommand].
     */
    override fun finishTopLevel() {
        finishTopLevel(null)
    }

    override fun finishTopLevel(animationData: AnimationData?) {
        executeCommand(FinishCommand(null, true, animationData))
    }

    /**
     * Finishes a current top-level screen (represented by activity) with result. Implemented with [FinishCommand].
     *
     * @param screenResult screen result that will be returned
     */
    override fun finishTopLevelWithResult(screenResult: ScreenResult) {
        finishTopLevelWithResult(screenResult, null)
    }

    override fun finishTopLevelWithResult(
        screenResult: ScreenResult,
        animationData: AnimationData?
    ) {
        executeCommand(FinishCommand(screenResult, true, animationData))
    }

    /**
     * Switches screen with a screen switcher. Implemented with [SwitchToCommand].
     *
     * @param screen screen
     */
    override fun switchTo(screen: Screen) {
        switchTo(screen, null)
    }

    override fun switchTo(screen: Screen, animationData: AnimationData?) {
        executeCommand(SwitchToCommand(screen, animationData))
    }

    private fun executeCommand(command: Command) {
        checkThatMainThread()
        commandQueue.add(command)
        executeQueuedCommands()
    }

    private fun executeQueuedCommands() {
        if (isExecutingCommands) {
            return
        }
        isExecutingCommands = true
        try {
            while (navigationContext != null && !commandQueue.isEmpty()) {
                val command = commandQueue.remove()
                val canExecuteCommands = command.execute(navigationContext!!)
                if (!canExecuteCommands) {
                    navigationContext = null
                }
            }
        } catch (e: NavigationException) {
            commandQueue.clear()
            navigationContext!!.navigationErrorListener.onNavigationError(e)
        } catch (e: Exception) {
            commandQueue.clear()
            throw e
        } finally {
            isExecutingCommands = false
        }
    }

    private fun checkThatMainThread() {
        if (Thread.currentThread() !== Looper.getMainLooper().thread) {
            throw RuntimeException("Can only be called from the main thread.")
        }
    }
}